cpp/src/arrow/python/python_to_arrow.cc

```diff
@@ -329,985 +303,649 @@ struct ValueConverter<DurationType> {
+  PyBytesView view_;
+  bool observed_binary_ = false;
 };
 
-// ----------------------------------------------------------------------
-// Convert lists (NumPy arrays containing lists or ndarrays as values)
-
-// If the value type does not match the expected NumPy dtype, then fall through
-// to a slower PySequence-based path
-#define LIST_FAST_CASE(TYPE, NUMPY_TYPE, ArrowType)            \
-  case Type::TYPE: {                                           \
-    if (PyArray_DESCR(arr)->type_num != NUMPY_TYPE) {          \
-      return value_converter_->Extend(obj, value_length);      \
-    }                                                          \
-    return AppendNdarrayTypedItem<NUMPY_TYPE, ArrowType>(arr); \
-  }
-
-// Use internal::VisitSequence, fast for NPY_OBJECT but slower otherwise
-#define LIST_SLOW_CASE(TYPE)                            \
-  case Type::TYPE: {                                    \
-    return value_converter_->Extend(obj, value_length); \
+template <typename U>
+class PyDictionaryConverter<U, enable_if_has_c_type<U>>
+    : public DictionaryConverter<U, PyConverter> {
+ public:
+  Status Append(PyObject* value) override {
+    if (PyValue::IsNull(this->options_, value)) {
+      return this->value_builder_->AppendNull();
+    } else {
+      ARROW_ASSIGN_OR_RAISE(auto converted,
+                            PyValue::Convert(this->value_type_, this->options_, value));
+      return this->value_builder_->Append(converted);
+    }
   }
+};
 
-// Base class for ListConverter and FixedSizeListConverter (to have both work with CRTP)
-template <typename TypeClass, NullCoding null_coding>
-class BaseListConverter : public TypedConverter<TypeClass, null_coding> {
+template <typename U>
+class PyDictionaryConverter<U, enable_if_has_string_view<U>>
+    : public DictionaryConverter<U, PyConverter> {
  public:
-  using BuilderType = typename TypeTraits<TypeClass>::BuilderType;
-
-  explicit BaseListConverter(bool from_pandas, bool strict_conversions,
-                             bool ignore_timezone)
-      : from_pandas_(from_pandas),
-        strict_conversions_(strict_conversions),
-        ignore_timezone_(ignore_timezone) {}
-
-  Status Init(ArrayBuilder* builder) override {
-    this->builder_ = builder;
-    this->typed_builder_ = checked_cast<BuilderType*>(builder);
-
-    this->value_type_ = checked_cast<const TypeClass&>(*builder->type()).value_type();
-    RETURN_NOT_OK(GetConverter(value_type_, from_pandas_, strict_conversions_,
-                               ignore_timezone_, &value_converter_));
-    return this->value_converter_->Init(this->typed_builder_->value_builder());
+  Status Append(PyObject* value) override {
+    if (PyValue::IsNull(this->options_, value)) {
+      return this->value_builder_->AppendNull();
+    } else {
+      ARROW_RETURN_NOT_OK(
+          PyValue::Convert(this->value_type_, this->options_, value, view_));
+      return this->value_builder_->Append(view_.bytes, view_.size);
```

Comment:
You're getting a conversion error here, cast to the expected integer type explicitly
https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/35406728/job/8kbxrx77m3x6mfjy#L1069